### PR TITLE
FractionalMaxPool3d with no_batch_dim support

### PR DESCRIPTION
**Pull request by george-qi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69732

Differential Revision: [D33280090](https://our.internmc.facebook.com/intern/diff/D33280090)

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/functional/pooling.h [hunk 1]
```diff
@@ -849,14 +849,15 @@ inline std::tuple<Tensor, Tensor> fractional_max_pool3d_with_indices(
...
-                    (int64_t)(input.sizes()[4] * (*output_ratio.value())[2])};
+    output_size_ = {(int64_t)(input.sizes()[-3] * (*output_ratio.value())[0]),
+                    (int64_t)(input.sizes()[-2] * (*output_ratio.value())[1]),
+                    (int64_t)(input.sizes()[-1] * (*output_ratio.value())[2])};
   }
 
   Tensor _random_samples_ = _random_samples;
   if (!_random_samples_.defined()) {
-    _random_samples_ = torch::rand({input.size(0), input.size(1), 3}, torch::TensorOptions().dtype(input.dtype()).device(input.device()));
+    auto n_batch = 1 ? input.dim() == 4 : input.size(0);
```
Comment (jbschlosser) [inline torch/csrc/api/include/torch/nn/functional/pooling.h:859]: I think you want `auto n_batch = input.dim() == 4 ? 1 : input.size(0)`, no?

File: torch/csrc/api/include/torch/nn/functional/pooling.h [hunk 2]
```diff
@@ -849,14 +849,15 @@ inline std::tuple<Tensor, Tensor> fractional_max_pool3d_with_indices(
   c10::optional<ExpandingArray<3>> output_size_ = output_size;
   if (output_size_ == c10::nullopt) {
     TORCH_INTERNAL_ASSERT(output_ratio != c10::nullopt);
-    output_size_ = {(int64_t)(input.sizes()[2] * (*output_ratio.value())[0]),
-                    (int64_t)(input.sizes()[3] * (*output_ratio.value())[1]),
-                    (int64_t)(input.sizes()[4] * (*output_ratio.value())[2])};
+    output_size_ = {(int64_t)(input.sizes()[-3] * (*output_ratio.value())[0]),
+                    (int64_t)(input.sizes()[-2] * (*output_ratio.value())[1]),
+                    (int64_t)(input.sizes()[-1] * (*output_ratio.value())[2])};
```
Comment (zou3519) [inline torch/csrc/api/include/torch/nn/functional/pooling.h:854]: Indexing into IntArrayRef doesn't wrap around. input.sizes()[-1] actually tries to index `input.sizes()[MAX_INT - 1]` which is definitely not what you wanted

Comment (zou3519) [reply torch/csrc/api/include/torch/nn/functional/pooling.h:854]: input.size(-1) should do the trick though

**Comment by george-qi** [comment]
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33280090).

File: torch/csrc/api/include/torch/nn/functional/pooling.h [hunk 3]
```diff
@@ -4,6 +4,7 @@
 #include <torch/nn/functional/activation.h>
 #include <torch/nn/options/pooling.h>
 #include <torch/nn/modules/utils.h>
+#include <iostream>
```
Comment (jbschlosser) [inline torch/csrc/api/include/torch/nn/functional/pooling.h:7]: nit: was this added for debugging? if so, it should be removed

**Review by jbschlosser — COMMENTED** [review]
Looking good! Just one tiny nit

**Comment by george-qi** [comment]
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33280090).

**Review by jbschlosser — APPROVED** [review]
LGTM!

**Pull request closed without merging**